File: src/conventions/pipeline/alert.ts
import type { BidAlert } from "../../core/contracts/bidding";
import type { BidMeaningClause } from "../../core/contracts/meaning";

// ─── Disclosure-based conventionality derivation ─────────────────────
//
// Alertability is derived from the `disclosure` field on BidMeaning.
// Constraints carry `isPublic` on the FactConstraint itself; no separate
// publicConstraints pipeline is needed.

/** Minimal shape for alert resolution — only needs disclosure and teachingLabel.
 *  Alertability is derived from the `disclosure` field. */
export interface AlertResolvable {
  readonly disclosure: "alert" | "announcement" | "natural" | "standard";
  readonly clauses: readonly BidMeaningClause[];
  readonly teachingLabel: string;
}

/** Derive whether a bid is alertable from its disclosure classification.
 *  A bid is alertable (requires some form of opponent notification) unless
 *  it has natural meaning (disclosure === "natural"). */
export function isAlertable(
  disclosure: "alert" | "announcement" | "natural" | "standard",
): boolean {
  return disclosure !== "natural";
}

/** Resolve whether a surface is alertable.
 *  Returns BidAlert when the bid is conventional, null when natural.
 *  Alertability is derived from the `disclosure` field.
 *  annotationType distinguishes ACBL alerts, announcements, and educational labels. */
export function resolveAlert(surface: AlertResolvable): BidAlert | null {
  if (!isAlertable(surface.disclosure)) return null;

  const annotationType = surface.disclosure === "announcement"
    ? "announce" as const
    : surface.disclosure === "standard"
      ? "educational" as const
      : "alert" as const;

  return {
    teachingLabel: surface.teachingLabel,
    annotationType,
  };
}

File: src-tauri/crates/bridge-conventions/src/pipeline/evaluation/alert.rs
//! Alert/disclosure resolution.
//!
//! Mirrors TS from `pipeline/evaluation/alert.ts`.

use serde::{Deserialize, Serialize};

use crate::types::meaning::Disclosure;

/// How a bid should be annotated for opponents.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum AnnotationType {
    Alert,
    Announce,
    Educational,
}

/// Alert resolution result.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BidAlert {
    pub alertable: bool,
    pub annotation_type: Option<AnnotationType>,
}

/// Check if a disclosure level requires opponent notification.
pub fn is_alertable(disclosure: Disclosure) -> bool {
    matches!(disclosure, Disclosure::Alert | Disclosure::Announcement)
}

/// Resolve a BidAlert from a disclosure level.
pub fn resolve_alert(disclosure: Disclosure) -> BidAlert {
    match disclosure {
        Disclosure::Alert => BidAlert {
            alertable: true,
            annotation_type: Some(AnnotationType::Alert),
        },
        Disclosure::Announcement => BidAlert {
            alertable: true,
            annotation_type: Some(AnnotationType::Announce),
        },
        Disclosure::Natural | Disclosure::Standard => BidAlert {
            alertable: false,
            annotation_type: Some(AnnotationType::Educational),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn alert_is_alertable() {
        assert!(is_alertable(Disclosure::Alert));
        assert!(is_alertable(Disclosure::Announcement));
        assert!(!is_alertable(Disclosure::Natural));
        assert!(!is_alertable(Disclosure::Standard));
    }
}
